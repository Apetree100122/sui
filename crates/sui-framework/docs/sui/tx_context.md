---
title: Module `sui::tx_context`
---



-  [Struct `TxContext`](#sui_tx_context_TxContext)
-  [Constants](#@Constants_0)
-  [Function `sender`](#sui_tx_context_sender)
-  [Function `digest`](#sui_tx_context_digest)
-  [Function `epoch`](#sui_tx_context_epoch)
-  [Function `epoch_timestamp_ms`](#sui_tx_context_epoch_timestamp_ms)
-  [Function `sponsor`](#sui_tx_context_sponsor)
-  [Function `fresh_object_address`](#sui_tx_context_fresh_object_address)
-  [Function `increment_ids_created`](#sui_tx_context_increment_ids_created)
-  [Function `ids_created`](#sui_tx_context_ids_created)
-  [Function `derive_id`](#sui_tx_context_derive_id)
-  [Function `native_sender`](#sui_tx_context_native_sender)
-  [Function `native_digest`](#sui_tx_context_native_digest)
-  [Function `native_epoch`](#sui_tx_context_native_epoch)
-  [Function `native_epoch_timestamp_ms`](#sui_tx_context_native_epoch_timestamp_ms)
-  [Function `native_sponsor`](#sui_tx_context_native_sponsor)
-  [Function `native_ids_created`](#sui_tx_context_native_ids_created)
-  [Function `native_inc_ids_created`](#sui_tx_context_native_inc_ids_created)


<pre><code><b>use</b> <a href="../std/option.md#std_option">std::option</a>;
<b>use</b> <a href="../std/vector.md#std_vector">std::vector</a>;
</code></pre>



<a name="sui_tx_context_TxContext"></a>

## Struct `TxContext`

Information about the transaction currently being executed.
This cannot be constructed by a transaction--it is a privileged object created by
the VM and passed in to the entrypoint of the transaction as <code>&<b>mut</b> <a href="../sui/tx_context.md#sui_tx_context_TxContext">TxContext</a></code>.


<pre><code><b>public</b> <b>struct</b> <a href="../sui/tx_context.md#sui_tx_context_TxContext">TxContext</a> <b>has</b> drop
</code></pre>



<details>
<summary>Fields</summary>


<dl>
<dt>
<code><a href="../sui/tx_context.md#sui_tx_context_sender">sender</a>: <b>address</b></code>
</dt>
<dd>
 The address of the user that signed the current transaction
</dd>
<dt>
<code>tx_hash: vector&lt;u8&gt;</code>
</dt>
<dd>
 Hash of the current transaction
</dd>
<dt>
<code><a href="../sui/tx_context.md#sui_tx_context_epoch">epoch</a>: u64</code>
</dt>
<dd>
 The current epoch number
</dd>
<dt>
<code><a href="../sui/tx_context.md#sui_tx_context_epoch_timestamp_ms">epoch_timestamp_ms</a>: u64</code>
</dt>
<dd>
 Timestamp that the epoch started at
</dd>
<dt>
<code><a href="../sui/tx_context.md#sui_tx_context_ids_created">ids_created</a>: u64</code>
</dt>
<dd>
 Counter recording the number of fresh id's created while executing
 this transaction. Always 0 at the start of a transaction
</dd>
</dl>


</details>

<a name="@Constants_0"></a>

## Constants


<a name="sui_tx_context_EUnsupportedFunction"></a>



<pre><code><b>const</b> <a href="../sui/tx_context.md#sui_tx_context_EUnsupportedFunction">EUnsupportedFunction</a>: u64 = 2;
</code></pre>



<a name="sui_tx_context_NATIVE_CONTEXT"></a>



<pre><code><b>const</b> <a href="../sui/tx_context.md#sui_tx_context_NATIVE_CONTEXT">NATIVE_CONTEXT</a>: bool = <b>false</b>;
</code></pre>



<a name="sui_tx_context_sender"></a>

## Function `sender`

Return the address of the user that signed the current
transaction


<pre><code><b>public</b> <b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_sender">sender</a>(self: &<a href="../sui/tx_context.md#sui_tx_context_TxContext">sui::tx_context::TxContext</a>): <b>address</b>
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_sender">sender</a>(self: &<a href="../sui/tx_context.md#sui_tx_context_TxContext">TxContext</a>): <b>address</b> {
    <b>if</b> (<a href="../sui/tx_context.md#sui_tx_context_NATIVE_CONTEXT">NATIVE_CONTEXT</a>) {
        <a href="../sui/tx_context.md#sui_tx_context_native_sender">native_sender</a>()
    } <b>else</b> {
        self.<a href="../sui/tx_context.md#sui_tx_context_sender">sender</a>
    }
}
</code></pre>



</details>

<a name="sui_tx_context_digest"></a>

## Function `digest`

Return the transaction digest (hash of transaction inputs).
Please do not use as a source of randomness.


<pre><code><b>public</b> <b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_digest">digest</a>(self: &<a href="../sui/tx_context.md#sui_tx_context_TxContext">sui::tx_context::TxContext</a>): &vector&lt;u8&gt;
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_digest">digest</a>(self: &<a href="../sui/tx_context.md#sui_tx_context_TxContext">TxContext</a>): &vector&lt;u8&gt; {
    <b>if</b> (<a href="../sui/tx_context.md#sui_tx_context_NATIVE_CONTEXT">NATIVE_CONTEXT</a>) {
        <a href="../sui/tx_context.md#sui_tx_context_native_digest">native_digest</a>()
    } <b>else</b> {
        &self.tx_hash
    }
}
</code></pre>



</details>

<a name="sui_tx_context_epoch"></a>

## Function `epoch`

Return the current epoch


<pre><code><b>public</b> <b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_epoch">epoch</a>(self: &<a href="../sui/tx_context.md#sui_tx_context_TxContext">sui::tx_context::TxContext</a>): u64
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_epoch">epoch</a>(self: &<a href="../sui/tx_context.md#sui_tx_context_TxContext">TxContext</a>): u64 {
    <b>if</b> (<a href="../sui/tx_context.md#sui_tx_context_NATIVE_CONTEXT">NATIVE_CONTEXT</a>) {
        <a href="../sui/tx_context.md#sui_tx_context_native_epoch">native_epoch</a>()
    } <b>else</b> {
        self.<a href="../sui/tx_context.md#sui_tx_context_epoch">epoch</a>
    }
}
</code></pre>



</details>

<a name="sui_tx_context_epoch_timestamp_ms"></a>

## Function `epoch_timestamp_ms`

Return the epoch start time as a unix timestamp in milliseconds.


<pre><code><b>public</b> <b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_epoch_timestamp_ms">epoch_timestamp_ms</a>(self: &<a href="../sui/tx_context.md#sui_tx_context_TxContext">sui::tx_context::TxContext</a>): u64
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_epoch_timestamp_ms">epoch_timestamp_ms</a>(self: &<a href="../sui/tx_context.md#sui_tx_context_TxContext">TxContext</a>): u64 {
    <b>if</b> (<a href="../sui/tx_context.md#sui_tx_context_NATIVE_CONTEXT">NATIVE_CONTEXT</a>) {
        <a href="../sui/tx_context.md#sui_tx_context_native_epoch_timestamp_ms">native_epoch_timestamp_ms</a>()
    } <b>else</b> {
        self.<a href="../sui/tx_context.md#sui_tx_context_epoch_timestamp_ms">epoch_timestamp_ms</a>
    }
}
</code></pre>



</details>

<a name="sui_tx_context_sponsor"></a>

## Function `sponsor`



<pre><code><b>public</b> <b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_sponsor">sponsor</a>(_self: &<a href="../sui/tx_context.md#sui_tx_context_TxContext">sui::tx_context::TxContext</a>): <a href="../std/option.md#std_option_Option">std::option::Option</a>&lt;<b>address</b>&gt;
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_sponsor">sponsor</a>(_self: &<a href="../sui/tx_context.md#sui_tx_context_TxContext">TxContext</a>): Option&lt;<b>address</b>&gt; {
    <b>assert</b>!(<a href="../sui/tx_context.md#sui_tx_context_NATIVE_CONTEXT">NATIVE_CONTEXT</a>, <a href="../sui/tx_context.md#sui_tx_context_EUnsupportedFunction">EUnsupportedFunction</a>);
    <a href="../sui/tx_context.md#sui_tx_context_native_sponsor">native_sponsor</a>()
}
</code></pre>



</details>

<a name="sui_tx_context_fresh_object_address"></a>

## Function `fresh_object_address`

Create an <code><b>address</b></code> that has not been used. As it is an object address, it will never
occur as the address for a user.
In other words, the generated address is a globally unique object ID.


<pre><code><b>public</b> <b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_fresh_object_address">fresh_object_address</a>(ctx: &<b>mut</b> <a href="../sui/tx_context.md#sui_tx_context_TxContext">sui::tx_context::TxContext</a>): <b>address</b>
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_fresh_object_address">fresh_object_address</a>(ctx: &<b>mut</b> <a href="../sui/tx_context.md#sui_tx_context_TxContext">TxContext</a>): <b>address</b> {
    <b>let</b> <a href="../sui/tx_context.md#sui_tx_context_ids_created">ids_created</a> = ctx.<a href="../sui/tx_context.md#sui_tx_context_ids_created">ids_created</a>();
    <b>let</b> id = <a href="../sui/tx_context.md#sui_tx_context_derive_id">derive_id</a>(*ctx.<a href="../sui/tx_context.md#sui_tx_context_digest">digest</a>(), <a href="../sui/tx_context.md#sui_tx_context_ids_created">ids_created</a>);
    ctx.<a href="../sui/tx_context.md#sui_tx_context_increment_ids_created">increment_ids_created</a>();
    id
}
</code></pre>



</details>

<a name="sui_tx_context_increment_ids_created"></a>

## Function `increment_ids_created`



<pre><code><b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_increment_ids_created">increment_ids_created</a>(self: &<b>mut</b> <a href="../sui/tx_context.md#sui_tx_context_TxContext">sui::tx_context::TxContext</a>)
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_increment_ids_created">increment_ids_created</a>(self: &<b>mut</b> <a href="../sui/tx_context.md#sui_tx_context_TxContext">TxContext</a>) {
    <b>if</b> (<a href="../sui/tx_context.md#sui_tx_context_NATIVE_CONTEXT">NATIVE_CONTEXT</a>) {
        self.<a href="../sui/tx_context.md#sui_tx_context_native_inc_ids_created">native_inc_ids_created</a>()
    } <b>else</b> {
        self.<a href="../sui/tx_context.md#sui_tx_context_ids_created">ids_created</a> = self.<a href="../sui/tx_context.md#sui_tx_context_ids_created">ids_created</a> + 1
    }
}
</code></pre>



</details>

<a name="sui_tx_context_ids_created"></a>

## Function `ids_created`

Return the number of id's created by the current transaction.
Hidden for now, but may expose later


<pre><code><b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_ids_created">ids_created</a>(self: &<a href="../sui/tx_context.md#sui_tx_context_TxContext">sui::tx_context::TxContext</a>): u64
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_ids_created">ids_created</a>(self: &<a href="../sui/tx_context.md#sui_tx_context_TxContext">TxContext</a>): u64 {
    <b>if</b> (<a href="../sui/tx_context.md#sui_tx_context_NATIVE_CONTEXT">NATIVE_CONTEXT</a>) {
        <a href="../sui/tx_context.md#sui_tx_context_native_ids_created">native_ids_created</a>()
    } <b>else</b> {
        self.<a href="../sui/tx_context.md#sui_tx_context_ids_created">ids_created</a>
    }
}
</code></pre>



</details>

<a name="sui_tx_context_derive_id"></a>

## Function `derive_id`

Native function for deriving an ID via hash(tx_hash || ids_created)


<pre><code><b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_derive_id">derive_id</a>(tx_hash: vector&lt;u8&gt;, <a href="../sui/tx_context.md#sui_tx_context_ids_created">ids_created</a>: u64): <b>address</b>
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>native</b> <b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_derive_id">derive_id</a>(tx_hash: vector&lt;u8&gt;, <a href="../sui/tx_context.md#sui_tx_context_ids_created">ids_created</a>: u64): <b>address</b>;
</code></pre>



</details>

<a name="sui_tx_context_native_sender"></a>

## Function `native_sender`



<pre><code><b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_native_sender">native_sender</a>(): <b>address</b>
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>native</b> <b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_native_sender">native_sender</a>(): <b>address</b>;
</code></pre>



</details>

<a name="sui_tx_context_native_digest"></a>

## Function `native_digest`



<pre><code><b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_native_digest">native_digest</a>(): &vector&lt;u8&gt;
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>native</b> <b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_native_digest">native_digest</a>(): &vector&lt;u8&gt;;
</code></pre>



</details>

<a name="sui_tx_context_native_epoch"></a>

## Function `native_epoch`



<pre><code><b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_native_epoch">native_epoch</a>(): u64
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>native</b> <b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_native_epoch">native_epoch</a>(): u64;
</code></pre>



</details>

<a name="sui_tx_context_native_epoch_timestamp_ms"></a>

## Function `native_epoch_timestamp_ms`



<pre><code><b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_native_epoch_timestamp_ms">native_epoch_timestamp_ms</a>(): u64
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>native</b> <b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_native_epoch_timestamp_ms">native_epoch_timestamp_ms</a>(): u64;
</code></pre>



</details>

<a name="sui_tx_context_native_sponsor"></a>

## Function `native_sponsor`



<pre><code><b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_native_sponsor">native_sponsor</a>(): <a href="../std/option.md#std_option_Option">std::option::Option</a>&lt;<b>address</b>&gt;
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>native</b> <b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_native_sponsor">native_sponsor</a>(): Option&lt;<b>address</b>&gt;;
</code></pre>



</details>

<a name="sui_tx_context_native_ids_created"></a>

## Function `native_ids_created`



<pre><code><b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_native_ids_created">native_ids_created</a>(): u64
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>native</b> <b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_native_ids_created">native_ids_created</a>(): u64;
</code></pre>



</details>

<a name="sui_tx_context_native_inc_ids_created"></a>

## Function `native_inc_ids_created`



<pre><code><b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_native_inc_ids_created">native_inc_ids_created</a>(self: &<b>mut</b> <a href="../sui/tx_context.md#sui_tx_context_TxContext">sui::tx_context::TxContext</a>)
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>native</b> <b>fun</b> <a href="../sui/tx_context.md#sui_tx_context_native_inc_ids_created">native_inc_ids_created</a>(self: &<b>mut</b> <a href="../sui/tx_context.md#sui_tx_context_TxContext">TxContext</a>);
</code></pre>



</details>
