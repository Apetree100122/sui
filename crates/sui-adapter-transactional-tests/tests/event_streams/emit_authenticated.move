// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

// Exercise test functions that emit authenticated events

//# init --addresses test=0x0 --accounts A B --simulator

//# publish

module test::event_streams {
    use sui::event;

    public struct Object has key, store {
        id: UID,
        event_stream: event::EventStream,
        event_stream_cap: event::EventStreamCap,
    }

    public struct AuthEvent has copy, drop {
        value: u64
    }

    public entry fun create(recipient: address, ctx: &mut TxContext) {
        let stream = event::new_event_stream(ctx);
        let cap = stream.get_cap(ctx);
        transfer::public_transfer(
            Object {
                id: object::new(ctx),
                event_stream: stream,
                event_stream_cap: cap,
            },
            recipient
        )
    }

    // test that reading o2 and updating o1 works
    public entry fun emit_event(o: &mut Object, value: u64) {
        // emit an event so the world can see the new value
        let ev = AuthEvent { value };
        o.event_stream_cap.emit(ev);
    }

    public entry fun delete(o: Object) {
        let Object { id, event_stream, event_stream_cap } = o;
        event_stream_cap.destroy();
        event_stream.destroy();
        object::delete(id);
    }
}

//# run test::event_streams::create --sender A --args @A

//# view-object 2,0

// TODO(deepak): you can add more event emits here for testing
// Run the test with:
// $ cargo nextest run -p sui-adapter-transactional-tests emit_authenticated

//# run test::event_streams::emit_event --sender A --args object(2,0) 10

//# create-checkpoint

//# view-object 4,0

//# run test::event_streams::delete --sender A --args object(2,0)
