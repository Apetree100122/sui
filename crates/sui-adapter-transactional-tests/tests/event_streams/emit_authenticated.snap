---
source: external-crates/move/crates/move-transactional-test-runner/src/framework.rs
---
processed 8 tasks

init:
A: object(0,0), B: object(0,1)

task 1, lines 8-49:
//# publish
created: object(1,0)
mutated: object(0,2)
gas summary: computation_cost: 1000000, storage_cost: 8420800,  storage_rebate: 0, non_refundable_storage_fee: 0

task 2, line 51:
//# run test::event_streams::create --sender A --args @A
created: object(2,0)
mutated: object(0,0)
gas summary: computation_cost: 1000000, storage_cost: 3055200,  storage_rebate: 0, non_refundable_storage_fee: 0

task 3, lines 53-57:
//# view-object 2,0
Owner: Account Address ( A )
Version: 2
Contents: test::event_streams::Object {
    id: sui::object::UID {
        id: sui::object::ID {
            bytes: fake(2,0),
        },
    },
    event_stream: sui::event::EventStream {
        name: sui::object::UID {
            id: sui::object::ID {
                bytes: _,
            },
        },
    },
    event_stream_cap: sui::event::EventStreamCap {
        id: sui::object::UID {
            id: sui::object::ID {
                bytes: _,
            },
        },
        stream_id: _,
    },
}

task 4, line 59:
//# run test::event_streams::emit_event --sender A --args object(2,0) 10
events: Event { package_id: test, transaction_module: Identifier("event_streams"), sender: A, type_: StructTag { address: test, module: Identifier("event_streams"), name: Identifier("AuthEvent"), type_params: [] }, contents: [10, 0, 0, 0, 0, 0, 0, 0] }
mutated: object(0,0), object(2,0)
accumulators_written: fake(4,0)
gas summary: computation_cost: 1000000, storage_cost: 3055200,  storage_rebate: 3024648, non_refundable_storage_fee: 30552

task 5, line 61:
//# create-checkpoint
Checkpoint created: 1

task 6, line 63:
//# view-object 4,0
Owner: Object ID: ( _ )
Version: 1
Contents: sui::dynamic_field::Field<sui::accumulator::Key, sui::event::EventStreamHead> {
    id: sui::object::UID {
        id: sui::object::ID {
            bytes: fake(4,0),
        },
    },
    name: sui::accumulator::Key {
        address: _,
        ty: vector[
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            48u8,
            50u8,
            58u8,
            58u8,
            101u8,
            118u8,
            101u8,
            110u8,
            116u8,
            58u8,
            58u8,
            69u8,
            118u8,
            101u8,
            110u8,
            116u8,
            83u8,
            116u8,
            114u8,
            101u8,
            97u8,
            109u8,
            72u8,
            101u8,
            97u8,
            100u8,
        ],
    },
    value: sui::event::EventStreamHead {
        root: vector[
            248u8,
            151u8,
            166u8,
            247u8,
            255u8,
            3u8,
            232u8,
            162u8,
            193u8,
            4u8,
            123u8,
            10u8,
            23u8,
            105u8,
            48u8,
            249u8,
            102u8,
            79u8,
            43u8,
            231u8,
            157u8,
            253u8,
            176u8,
            96u8,
            4u8,
            83u8,
            199u8,
            173u8,
            175u8,
            74u8,
            252u8,
            124u8,
        ],
        prev: vector[
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
            0u8,
        ],
    },
}

task 7, line 65:
//# run test::event_streams::delete --sender A --args object(2,0)
mutated: object(0,0)
deleted: object(2,0)
unwrapped_then_deleted: object(_), object(_)
gas summary: computation_cost: 1000000, storage_cost: 988000,  storage_rebate: 3024648, non_refundable_storage_fee: 30552
